### PR TITLE
Miscutil: Orcidpush improvements

### PR DESCRIPTION
**Pull request by MSusik**

### CODE REVIEW

**Comment by MSusik** [comment]
@kaplun  @glouppe

Sam, were you rebasing sth? This PR should contain only 2 commits, yet it still has two initial commits for Orcidpush.

File: modules/miscutil/lib/orcidutils.py [hunk 1]
```diff
@@ -0,0 +1,787 @@
...
+    if len(publication_array) > 0 and \
+            re.match(r'[12]\d{3}$', publication_array[0]):
+        result['year'] = publication_array[0]
+        if len(publication_array) > 1 and \
+                re.match(r'[01]\d$', publication_array[1]):
+            result['month'] = publication_array[1]
+            if len(publication_array) > 2 and \
+                    re.match(r'[012]\d{3}$', publication_array[2]):
+                result['day'] = publication_array[2]
+
```
Comment (MSusik) [inline modules/miscutil/lib/orcidutils.py:586]: This should be 'master proof'

Check for an example :

``` sql
select value from bib50x where tag = '502__c' order by value desc limit 100;
```

This field should indicate date :stuck_out_tongue_closed_eyes:

**Comment by kaplun** [comment]
What are you talking about? We have to talk :smile: But on Monday!

**Comment by MSusik** [comment]
I developed over the same branch that I used when I was writing orcidpush. Somehow, this PR contained also previous commits. It means that my previous commits were rebased (probably), so git didn't know that they were already included in prod. Anyway, I decided to rebase my new commits on head, so they won't cause any problems.

File: modules/miscutil/lib/orcidutils.py (hunk 1 above)
Comment (kaplun) [inline modules/miscutil/lib/orcidutils.py:586]: The official date of a publication is actually contained in `260__c` for published, and `269__c` for preprints. See:
https://twiki.cern.ch/twiki/bin/view/Inspire/DevelopmentRecordMarkup

`502__d` is the date of the degree (I guess valid only for thesis documents).

Comment (kaplun) [inline modules/miscutil/lib/orcidutils.py:586]: Ah but I see that indeed you checked all three values. Perfect.

**Comment by glouppe** [comment]
Even if this is ready, let just no merge this yet. From the last meeting on author quick wins, it was decided that this would require approval of everyone before being put into production.

**Comment by kaplun** [comment]
@MSusik can you check if it is possible to cleanly factor out the first commit, as it would be great to quickly deploy that on INSPIRE production, in order to fix the encoding bug.

**Comment by MSusik** [comment]
@kaplun I should implement the whole thing (20 years tokens) today or tomorrow, so I won't try to factor out anything. It will be better.

File: modules/miscutil/lib/orcidutils.py [hunk 2]
```diff
@@ -412,49 +415,61 @@ def _orcid_fetch_works(pid):
...
-        return False
-    if token == None:
-        bibtask.write_message("You should provide '--token' option!",
-                              stream=sys.stderr)
-        return False
+        pid = pid_token[0]
+        token = pid_token[1]
 
-    doid, xml_to_push = _orcid_fetch_works(int(pid))
+        if pid_token[2]:
```
Comment (MSusik) [inline modules/miscutil/lib/orcidutils.py:433]: So, the concept is easy: We store tokens in the aidTOKEN table. We also store there information whether we should push the data or not. Then, we iterate over every token, and we push the data where it is needed.

File: modules/miscutil/lib/orcidutils.py [hunk 3]
```diff
@@ -412,49 +415,61 @@ def _orcid_fetch_works(pid):
...
-            bibtask.write_message(response.text)
-        return False
+            response = requests.post(url, xml_to_push, headers=headers)
+
+            code = response.status_code
+
+            if code == 401:
+                # The token has expired or the user revoke his token
+                delete_token(pid)
+                bibtask.write_message("Token deleted for %s" % pid)
```
Comment (MSusik) [inline modules/miscutil/lib/orcidutils.py:461]: Whis means that the token expired. We delete the entry from the table and continue.

File: modules/bibauthorid/lib/bibauthorid_dbinterface.py [hunk 4]
```diff
@@ -4906,3 +4907,69 @@ def get_records_of_authors(personids_set):
...
+        try:
+            raise TokenExistingException
+        except:
+            register_exception(subject="""The token in aidTOKEN was updated.
+                                          This behaviour was not expected.""")
+    else:
+        # default behaviour
+        change = 1
+        run_sql("insert into aidTOKEN values \
+                (%s, %s, %s)", (personid, token, change))
```
Comment (MSusik) [inline modules/bibauthorid/lib/bibauthorid_dbinterface.py:4949]: If the token was inserted, it means that we need to push the works to ORCiD.

File: modules/bibauthorid/lib/bibauthorid_webapi.py [hunk 5]
```diff
@@ -3519,6 +3520,7 @@ def commit_ticket_guest(ticket, userinfo, uid, modified_pids):
     def commit_ticket_user(ticket, userinfo, uid, modified_pids):
         ok_ops = list()
         for op in list(ticket):
+            trigger_aidtoken_change(op['pid'], 1)
```
Comment (MSusik) [inline modules/bibauthorid/lib/bibauthorid_webapi.py:3523]: Here  we have second source of changes: the claims.

File: modules/bibauthorid/lib/bibauthorid_dbinterface.py [hunk 6]
```diff
@@ -4906,3 +4907,69 @@ def get_records_of_authors(personids_set):
...
+    """Trigger the change in aidTOKEN table on."""
+    run_sql("update aidTOKEN set `change`=%s where personid=%s",
+            (change, personid))
+
+# ********** getters **********#
+
+
+def get_token(personid):
+    """Get the ORCiD token for an author."""
+    return run_sql("select * from aidTOKEN where personid=%s", (personid,))
```
Comment (kaplun) [inline modules/bibauthorid/lib/bibauthorid_dbinterface.py]: Do not use `select *`: always specify the list of desired columns. In this way, in case we change the table structure later this query is going to still be robust.

See also https://github.com/tiborsimko/invenio-devscripts#invenio-check-kwalitee

File: modules/bibauthorid/lib/bibauthorid_dbinterface.py [hunk 7]
```diff
@@ -4906,3 +4907,69 @@ def get_records_of_authors(personids_set):
...
+
+
+def get_all_tokens():
+    """Get the ORCiD token for every author."""
+    return run_sql("select * from aidTOKEN")
```
Comment (kaplun) [inline modules/bibauthorid/lib/bibauthorid_dbinterface.py]: Ditto

File: modules/bibauthorid/lib/bibauthorid_templates.py [hunk 8]
```diff
@@ -2121,42 +2121,42 @@ def tmpl_orcid_box(self, orcid_data, ln, orcid_info, add_box=True, loading=True)
...
+                      <div class="modal-header"> \
+                      <h4 class="modal-title">%s</h4> \
+                      </div> \
+                      <div class="modal-body"> \
+                      <p>%s</p></div> \
+                      <div class="modal-footer"> \
+                      <button type="button" class="btn btn-default" data-dismiss="modal">%s</button> \
+                      <a rel="nofollow" href="%s" class="btn btn-primary">%s</a> \
+                      </div></div></div></div>' % (
+                        _("Pushing claimed publication list to ORCiD"),
```
Comment (kaplun) [inline modules/bibauthorid/lib/bibauthorid_templates.py]: I believe the correct modern spelling of _ORCID_ is all upper case.

File: modules/bibauthorid/lib/bibauthorid_webapi.py (hunk 5 above)
Comment (kaplun) [inline modules/bibauthorid/lib/bibauthorid_webapi.py:3523]: Aha. So you are triggering a re-push of an orcid, simply by touching the corresponding profile ID change flag. This could be error prone if we forget to amend the flag somewhere in the code.

Isn't it possible to decide which profiles have been updated by querying the `aidPERSONIDPAPERS` for all the `personid` having `last_updated` equal or grater to the last time `orcidpush` has run? (filtering for only those `personid` having an ORCID)

File: modules/bibauthorid/lib/bibauthorid_webinterface.py [hunk 9]
```diff
@@ -2790,10 +2791,9 @@ def __call__(self, req, form):
             "orcid_message" : self._generate_orcid_message(req, ln)
         }
 
-        if 'orcid_info' in pinfo:
-            if pinfo['orcid_info']['status'] not in ('running',):
-                pinfo.pop('orcid_info', None)
-                session.dirty = True
+        if 'orcid_info' in session:
+            session.pop('orcid_info', None)
+            session.dirty = True
```
Comment (kaplun) [inline modules/bibauthorid/lib/bibauthorid_webinterface.py:2796]: Oh: why this was popped away in the first place?

File: modules/miscutil/lib/orcidutils.py (hunk 1 above)
Comment (kaplun) [inline modules/miscutil/lib/orcidutils.py:586]: Minus `502__c` should indeed be `502__d`

File: modules/miscutil/lib/orcidutils.py [hunk 10]
```diff
@@ -540,6 +555,68 @@ def _get_orcid_dictionaries(papers, personid, old_external_ids):
...
+                  the date from the field indicated by this field
+    @type field: string
+
+    @return: dictionary
+    @rtype: dict
+    '''
+
+    result = {}
+
+    publication_date = record_get_field_value(recstruct, field, '', '', 'c')
```
Comment (kaplun) [inline modules/miscutil/lib/orcidutils.py]: Ah it's here that you hardcode the infamous `c`...

File: modules/webaccess/lib/external_authentication_oauth2.py [hunk 11]
```diff
@@ -47,7 +47,7 @@ def __init_req(req):
         req.g['oauth2_debug_msg'] = ''
         req.g['oauth2_response'] = None
 
-    def auth_user(self, username, password, req=None):
+    def auth_user(self, username, password, req=None, authors=False):
```
Comment (kaplun) [inline modules/webaccess/lib/external_authentication_oauth2.py]: Shall we better give a more meaningful name to this variable? E.g. `oauth_token_only` and document why this is exists at all?

File: modules/miscutil/lib/upgrades/invenio_2014_10_14_new_aidTOKEN.py [hunk 12]
```diff
@@ -0,0 +1,43 @@
...
+    return "New bibauthorid ORCiD tokens table"
+
+
+def do_upgrade():
+    """Create the ORCiD token table."""
+    run_sql("""CREATE TABLE IF NOT EXISTS `aidTOKEN` (
+      `personid` BIGINT( 16 ) UNSIGNED PRIMARY
+      KEY NOT NULL, `token` VARCHAR( 255 ) NOT NULL,
+      `change` BOOL NOT NULL
+    ) ENGINE=MyISAM""")
```
Comment (kaplun) [inline modules/miscutil/lib/upgrades/invenio_2014_10_14_new_aidTOKEN.py:38]: Great that you are providing the upgrade script. Please, correspondingly enrich `tabcreate.sql` and `tabdrop.sql`. In the former you will also have to specify this very upgrade script (see the end of the file).

File: modules/bibauthorid/lib/bibauthorid_webapi.py (hunk 5 above)
Comment (MSusik) [inline modules/bibauthorid/lib/bibauthorid_webapi.py:3523]: > This could be error prone if we forget to amend the flag somewhere in the code.

I agree. However

>  having `last_updated` equal or grater to the last time `orcidpush` has run? 

This means pushing also for all the authors that changed anything different than claims. For example a Rabbit run will cause an Orcidpush, despite the fact that claims were not moved.

File: modules/bibauthorid/lib/bibauthorid_webinterface.py (hunk 9 above)
Comment (MSusik) [inline modules/bibauthorid/lib/bibauthorid_webinterface.py:2796]: If I remember correctly orcid_info contained two things: an id of bibtask and a status.

A status was a source determining the message shown to the user. I the status was different than 'running' (so, for example error) we were showing this message to user only once. After showing it (this function is responsible for generating the `Manage Your Profile` page) we should remove this information from session so that it won't be should anymore.

File: modules/miscutil/lib/orcidutils.py (hunk 10 above)
Comment (MSusik) [inline modules/miscutil/lib/orcidutils.py]: Well, yes, it was wrong. The code was implying that everytime we are looking for a date, we need to to acess `c` subfield. Even if that was true I should have not written it that way.

**Comment by MSusik** [comment]
Thanks for review, I will improve the code quality soon.

File: modules/miscutil/lib/orcidutils.py (hunk 1 above)
Comment (MSusik) [inline modules/miscutil/lib/orcidutils.py:586]: Now I saw that I didn't change it here. It is correctly implemented in #123

File: modules/bibauthorid/lib/bibauthorid_templates.py (hunk 8 above)
Comment (MSusik) [inline modules/bibauthorid/lib/bibauthorid_templates.py]: The spelling `ORCiD` has been already used in `Manage Profile` page, so I will keep to this convention.

File: modules/miscutil/lib/orcidutils.py (hunk 10 above)
Comment (MSusik) [inline modules/miscutil/lib/orcidutils.py]: Will be improved during rebasing after #123 is merged.

File: modules/bibauthorid/lib/bibauthorid_templates.py (hunk 8 above)
Comment (kaplun) [inline modules/bibauthorid/lib/bibauthorid_templates.py]: Or rather correct it everywhere. As far as I know it's really wrong (and ugly). (check also with Patricia and Laura) Additionally ORCID integration has not yet been announced for real, so we are still in time to correct spelling etc.

File: modules/webaccess/lib/external_authentication_oauth2.py (hunk 11 above)
Comment (MSusik) [inline modules/webaccess/lib/external_authentication_oauth2.py]: Done in last commit.

It is  Episode 1 Series 2 of 'Sed as The Only Programming Tool You Need'.

File: modules/bibauthorid/lib/bibauthorid_dbinterface.py [hunk 13]
```diff
@@ -4906,3 +4907,69 @@ def get_records_of_authors(personids_set):
...
+
+
+def get_token(personid):
+    """Get the ORCID token for an author."""
+    return run_sql("select personid, token, `change` from aidTOKEN where personid=%s", (personid,))
+
+
+def get_all_tokens():
+    """Get the ORCID token for every author."""
+    return run_sql("select personid, token, `change` from aidTOKEN")
```
Comment (glouppe) [inline modules/bibauthorid/lib/bibauthorid_dbinterface.py]: Can you enforce code consistency by not using backticks?

File: modules/bibauthorid/lib/bibauthorid_templates.py [hunk 14]
```diff
@@ -2112,60 +2112,60 @@ def tmpl_arxiv_box(self, arxiv_data, ln, add_box=True, loading=True):
...
+                      </div> \
+                      <div class="modal-body"> \
+                      <p>%s</p></div> \
+                      <div class="modal-footer"> \
+                      <button type="button" class="btn btn-default" data-dismiss="modal">%s</button> \
+                      <a rel="nofollow" href="%s" class="btn btn-primary">%s</a> \
+                      </div></div></div></div>' % (
+                        _("Pushing claimed publication list to ORCID"),
+                        _("We allow you to push your works to ORCID. \
+                           To do that, after clicking 'Push to ORCID' button, \
```
Comment (glouppe) [inline modules/bibauthorid/lib/bibauthorid_templates.py]: clicking "on the"

File: modules/bibauthorid/lib/bibauthorid_templates.py [hunk 15]
```diff
@@ -2112,60 +2112,60 @@ def tmpl_arxiv_box(self, arxiv_data, ln, add_box=True, loading=True):
...
+                           you will have to succesfully authorize \
+                           to the ORCID account mentioned the current page. \
+                           We will send the details of all the papers and \
+                           datasets you have claimed as yours in INSPIRE. \
+                           After doing this, if you claim any paper on INSPIRE, we \
+                           will push it to ORCID.<br><p>Please select the \
+                           'Allow this permission until I revoke it' option \
+                           in the ORCID login form.</br></p>"),
```
Comment (glouppe) [inline modules/bibauthorid/lib/bibauthorid_templates.py:2156]: I would rephrase the whole thing more simply as:

"To push your publications on ORCID, press the 'Push to ORCID' button and authorize Inspire to access your ORCID account. "

File: modules/bibauthorid/lib/bibauthorid_templates.py [hunk 16]
```diff
@@ -2207,26 +2207,19 @@ def tmpl_orcid_message(self, orcid_info, ln):
...
+                    profile.'))
         elif orcid_info == 'wrong_account':
             html = ('<div class="alert alert-danger" role="alert">%s</div>' % _('You \
-                    authenticated correctly to ORCID, but you are using a different \
-                    account than the one that is connected to your profile on INSPIRE. \
-                    We will not allow you push your works to a different account. \
-                    If you want to change your ORCID on your INSPIRE profile, \
-                    please contact our staff.'))
+                    authenticated correctly to ORCID, but you authenticated \
+                    to a different account that the one that is connectd \
```
Comment (glouppe) [inline modules/bibauthorid/lib/bibauthorid_templates.py]: connected

File: modules/bibauthorid/lib/bibauthorid_dbinterface.py (hunk 13 above)
Comment (MSusik) [inline modules/bibauthorid/lib/bibauthorid_dbinterface.py]: `change` is some kind of a keyword in mysql. I can only change backticks to single quotes if you want.

File: modules/bibauthorid/lib/bibauthorid_webinterface.py [hunk 17]
```diff
@@ -2879,52 +2856,68 @@ def import_orcid_pubs(self, req, form):
...
             # Authenticate
-            uinfo['pushorcid'] = True
+            session['pushorcid'] = True
             session.dirty = True
             redirect_to_url(req, "%s/youraccount/oauth2?provider=%s&scope=/orcid-works/create" % (CFG_SITE_SECURE_URL, 'orcid'))
 
-        #We expect user to have only one ORCID
-        assert(len(webapi.get_orcids_by_pid(pid)) == 1)
+        # We expect user to have only one ORCID
+        assert(len(webapi.get_orcids_by_pid(session['orcid_pid'])) == 1)
```
Comment (glouppe) [inline modules/bibauthorid/lib/bibauthorid_webinterface.py:2899]: `personid=0` has 9 ORCID already in the prod db. How was this possible in the first place?

Comment (MSusik) [inline modules/bibauthorid/lib/bibauthorid_webinterface.py:2899]: I added this line of the code a month ago. Before that there were no assertions.

Comment (glouppe) [inline modules/bibauthorid/lib/bibauthorid_webinterface.py:2899]: Yes, but my comment was more general than this particular line of code. We do have ORCID that are assigned to personid=0, how is this possible? Dont we have a bug somewhere?

File: modules/bibauthorid/lib/bibauthorid_dbinterface.py (hunk 13 above)
Comment (glouppe) [inline modules/bibauthorid/lib/bibauthorid_dbinterface.py]: Then maybe should not use a mysql keyword to name a column.

File: modules/bibauthorid/lib/bibauthorid_webinterface.py (hunk 9 above)
Comment (kaplun) [inline modules/bibauthorid/lib/bibauthorid_webinterface.py:2796]: Things will be so nice with Flask: http://flask.pocoo.org/docs/0.10/patterns/flashing/

File: modules/bibauthorid/lib/bibauthorid_dbinterface.py (hunk 13 above)
Comment (kaplun) [inline modules/bibauthorid/lib/bibauthorid_dbinterface.py]: Good point. `change` is also a bit vague in meaning. What could be a better column name?

File: modules/bibauthorid/lib/bibauthorid_templates.py (hunk 15 above)
Comment (kaplun) [inline modules/bibauthorid/lib/bibauthorid_templates.py:2156]: What about "To synchronize your publications with ORCID, press the 'Push to ORCID' button and authorize Inspire to access your ORCID account. "? Something for the communication/UI/UX team. @jmartinm

File: modules/bibauthorid/lib/bibauthorid_dbinterface.py (hunk 13 above)
Comment (MSusik) [inline modules/bibauthorid/lib/bibauthorid_dbinterface.py]: I will change it to `was_changed`

**Comment by MSusik** [comment]
Sunday, 26th, ORCID introduces new tokens in production. So, how is the testing going :laughing: :question:

**Comment by kaplun** [comment]
Can you rebase it once to prod? I think e.g. the first commit was already integrated and I don't want to break up things.

**Comment by kaplun** [comment]
Doesn't matter. I have done it.

**Comment by kaplun** [comment]
Superseeded by #129
